### PR TITLE
Migrates to data modules

### PR DESCRIPTION
**Pull request by kylebgorman**
This PR migrates us to use lightning's built-in notion of data modules. This is a sort of data "God class" (by design) which holds the information about file paths, file parsing, collation/batching/padding, and the index, and can be used more or less the same during training, validation, prediction, or testing.

By migrating to this we are much closer to being able to use Lightning (and Torch!) 2.0, particularly the automagical LightningCLI, and we get rid of a lot of boilerplate in the training and prediction scripts. (I also thinks this will help enormously with testing.) Thus this is relevant to #5 and #60.

Two big notes:

* I decided to stop fighting PyTorch Lightning and am submitting to use `--val` where I would much prefer `--dev`, since that's what it expects.
* There is now just one `Dataset`. I think that's a big improvement. Indeed there are no "two-flavors-of" anything anymore, since @bonham79's work on models vs. modules and this. 

(I am closing #91 in favor of this; that had a bug I could never squash so I started over and worked more incrementally.)

### CODE REVIEW

File: yoyodyne/predict.py [hunk 1]
```diff
@@ -23,51 +22,35 @@ def get_trainer_from_argparse_args(
...
-    return data.DataLoader(
-        dataset,
-        collate_fn=collator,
-        batch_size=batch_size,
-        num_workers=1,
+    separate_features = args.features_col != 0 and args.arch in [
+        "pointer_generator_lstm",
+        "transducer",
+    ]
+    # TODO(kbg): reuse index?
```
Comment (Adamits) [inline yoyodyne/predict.py:40]: If I understand correctly, yes. We need to reuse the index from training for this to work.

File: yoyodyne/predict.py [hunk 2]
```diff
@@ -99,34 +82,28 @@ def _mkdir(output: str) -> None:
...
+    datamodule: data.DataModule,
     output: str,
 ) -> None:
     """Predicts from the model.
 
     Args:
          trainer (pl.Trainer).
          model (pl.LightningModule).
-         loader (data.DataLoader).
+         dataomdule (data.DataModule).
```
Comment (Adamits) [inline yoyodyne/predict.py:94]: `dataomdule` --> `datamodule`

**Review by Adamits — CHANGES_REQUESTED** [review]
This looks good, left a few comments.

I am wondering about two main things, though:

1. It seems that we **do** still need to use the dataset directly sometimes, and we access it with e.g.:`datamodule.train_dataloader().dataset`. To be sure, this is not a separate initialization from what PTL does under the hood? 
2. If I am a user that wants to modify some dataset behavior in a way that is not easily supported from the CLI (say, I want a 4th column for some metadata available to my script, or I want to somehow change my data at runtime based on some other info), what is the process of updating the code now? I just want to make sure that we have not increased complexity too much from this perspective.

**Comment by kylebgorman** [comment]
>
>
>    1. It seems that we *do* still need to use the dataset directly
>    sometimes, and we access it with e.g.:
>    datamodule.train_dataloader().dataset. To be sure, this is not a
>    separate initialization from what PTL does under the hood?
>
> Yes, that's how I'd do it. It comes up once, in feeding data to transducer
EM.

File: README.md [hunk 3]
```diff
@@ -54,8 +54,33 @@ import yoyodyne
...
+Training is performed by the [`yoyodyne-train`](yoyodyne.train.py) script. One
+must specify the following required arguments:
+
+-   `--train`: path to TSV file containing training data
+-   `--val`: path to TSV file containing validation data
+-   `--experiment`: name of experiment (pick something unique)
+-   `--model_dir`: path for model metadata and checkpoints output during
+    training
+
+The user can also specify as well as various optional training and architectural
```
[the remark this answers is not in the document]
Comment (kylebgorman) [reply README.md:68]: Fixed.

File: tests/collator_test.py
[the remark this answers is not in the document]
Comment (kylebgorman) [reply tests/collator_test.py:1]: These tests make no sense anymore. All the relevant logic lives inside the data module. The collator tester just checks that we passed the separate features boolean property; the dataset tester was really a test of `get_dataset`, which is no longer necessary because I merged the two dataset classes.

File: yoyodyne/data/collators.py [hunk 4]
```diff
@@ -1,57 +1,30 @@
...
-from . import batches, datasets, defaults, util
+from .. import defaults, util
+from . import batches, datasets
 
 
 class LengthError(Exception):
     pass
 
 
+@dataclasses.dataclass
```
[the remark this answers is not in the document]
Comment (kylebgorman) [reply yoyodyne/data/collators.py:17]: I also was initially uncertain about this but I convinced myself that Python just sees it as an arbitrary callable and that it doesn't have any PyTorch magic in it. (Note it doesn't inherent from `nn.Module` or anything like that.)

File: yoyodyne/data/datamodules.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+    # Helpers.
+
+    @property
+    def paths(self) -> Iterator[str]:
+        if self.train is not None:
+            yield self.train
+        if self.val is not None:
+            yield self.val
+        if self.predict is not None:
+            yield self.predict
```
[the remark this answers is not in the document]
Comment (kylebgorman) [reply yoyodyne/data/datamodules.py:117]: It isn't supported now; you'll get a parsing error. Same with `--test`.

I think with LightningCLI it will be technically supported but no-op, which also seems fine. (We could make it "op" if you want at a later point.)

File: yoyodyne/data/datamodules.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+                f"Features vocabulary: {self.index.features_map.pprint()}"
+            )
+        if self.index.has_target:
+            util.log_info(
+                f"Target vocabulary: {self.index.target_map.pprint()}"
+            )
+
+    def write_index(self, model_dir: str, experiment: str) -> None:
+        """Writes the index."""
+        index_path = self.index.index_path(model_dir, experiment)
```
[the remark this answers is not in the document]
Comment (kylebgorman) [reply yoyodyne/data/datamodules.py:135]: How about instead I just inline the computation of that path? Done.

File: yoyodyne/data/datasets.py [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+            symbol_map (indexes.SymbolMap).
+
+        Yields:
+            List[str]: Decoded symbols.
+        """
+        for idx in indices.cpu().numpy():
+            yield [
+                symbol_map.symbol(c)
+                for c in idx
+                if c not in self.index.special_idx
```
[the remark this answers is not in the document]
Comment (kylebgorman) [reply yoyodyne/data/datasets.py:148]: Removed for YAGNI reasons. We only ever decode target at present.

File: yoyodyne/data/datasets.py [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+        """Decodes a source tensor.
+
+        Args:
+            indices (torch.Tensor): 2d tensor of indices.
+
+        Yields:
+            str: Decoded source strings.
+        """
+        for symbols in self._decode(indices, self.index.source_map):
+            yield self.parser.source_string(symbols)
```
[the remark this answers is not in the document]
Comment (kylebgorman) [reply yoyodyne/data/datasets.py:164]: What do you want it to say, exactly? It links to a 1-line method that does just that, so I'd like more direction what would make this clearer.

File: yoyodyne/predict.py [hunk 9]
```diff
@@ -99,34 +82,28 @@ def _mkdir(output: str) -> None:
...
          output (str).
-         target_sep (str).
     """
-    dataset = loader.dataset
-    target_sep = dataset.config.target_sep
     util.log_info(f"Writing to {output}")
     _mkdir(output)
+    decode_target = datamodule.predict_dataloader().dataset.decode_target
```
[the remark this answers is not in the document]
Comment (kylebgorman) [reply yoyodyne/predict.py:99]: I have modified it so it's only run once.

I suppose you could put those decode methods there though I don't see a big improvement either way.

[the remark this answers is not in the document]
Comment (Adamits) [reply yoyodyne/predict.py]: Ok, seems fine then!

File: yoyodyne/data/datamodules.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Adamits) [reply yoyodyne/data/datamodules.py:117]: Sounds good. If we decide we want them "op", then that sounds like a separate PR.

**Comment by kylebgorman** [comment]
> Only leftover comment is about reusing the train index when making predictions, for which you have left a TODO. But unless this index is loaded elsewhere, this will be an issue if the train and test set are do not have identical vocabularies. It seems like this can be handled in a separate issue if you like.

I wasn't sure at first but I convinced myself that you're right and it's important. 

Just putting this here since we are already talking about it: after migration to LightningCLI, this will compute the index across any of `--train`, `--val`, `--predict`, `--test` as provided, independent of what command (training, prediction, etc.) is requested.

File: yoyodyne/predict.py (hunk 1 above)
Comment (kylebgorman) [reply yoyodyne/predict.py]: Done.
